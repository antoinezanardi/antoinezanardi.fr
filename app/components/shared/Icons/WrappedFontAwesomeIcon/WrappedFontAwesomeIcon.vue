<template>
  <span>
    <FontAwesomeIcon
      class="font-awesome-icon"
      :class="classes"
      :icon="icon"
      role="button"
      :size="size"
    />
  </span>
</template>

<script lang="ts" setup>
// TODO: exists because when building the app with Nuxt as static, FontAwesomeIcon duplicates its neighbor element in the generated HTML.
// it should be removed when fixed https://github.com/FortAwesome/vue-fontawesome/issues/384
import { FontAwesomeIcon } from "@fortawesome/vue-fontawesome";

import type { WrappedFontAwesomeIconProps } from "~/components/shared/Icons/WrappedFontAwesomeIcon/wrapped-font-awesome-icon.types";

withDefaults(defineProps<WrappedFontAwesomeIconProps>(), {
  iconColor: "#000000",
  size: undefined,
  classes: "",
});
</script>

<style lang="scss" scoped>
  .font-awesome-icon {
    color: v-bind(iconColor);
  }
</style>