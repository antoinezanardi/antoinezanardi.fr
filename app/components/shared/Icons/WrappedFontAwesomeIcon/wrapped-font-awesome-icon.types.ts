type WrappedFontAwesomeIconProps = {
  icon: string;
  iconColor?: string;
  size?: "1x" | "2x" | "3x" | "4x" | "5x" | "6x" | "7x" | "8x" | "9x" | "10x" | "lg" | "sm" | "xs";
  classes?: string;
};

export type { WrappedFontAwesomeIconProps };