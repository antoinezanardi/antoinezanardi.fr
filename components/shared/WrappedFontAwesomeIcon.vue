<template>
  <span>
    <FontAwesomeIcon class="font-awesome-icon" :icon="icon" :size="size" :class="classes" role="button"/>
  </span>
</template>

<script lang="ts" setup>
// TODO: exists because when building the app with Nuxt as static, FontAwesomeIcon duplicates its neighbor element in the generated HTML.
// it should be removed when fixed https://github.com/FortAwesome/vue-fontawesome/issues/384
import { FontAwesomeIcon } from "@fortawesome/vue-fontawesome";

withDefaults(defineProps<{
  icon: string;
  iconColor?: string;
  size?: "1x" | "2x" | "3x" | "4x" | "5x" | "6x" | "7x" | "8x" | "9x" | "10x" | "lg" | "sm" | "xs";
  classes?: string;
}>(), {
  iconColor: "#000000",
  size: undefined,
  classes: "",
});
</script>

<style lang="scss" scoped>
  .font-awesome-icon {
    color: v-bind(iconColor);
  }
</style>