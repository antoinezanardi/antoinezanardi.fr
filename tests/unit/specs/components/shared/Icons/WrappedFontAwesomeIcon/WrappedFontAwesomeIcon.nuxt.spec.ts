import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";

import type { WrappedFontAwesomeIconProps } from "~/components/shared/Icons/WrappedFontAwesomeIcon/wrapped-font-awesome-icon.types";
import WrappedFontAwesomeIcon from "~/components/shared/Icons/WrappedFontAwesomeIcon/WrappedFontAwesomeIcon.vue";
import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";

describe("Wrapped Font Awesome Icon Component", () => {
  const defaultProps: WrappedFontAwesomeIconProps = {
    icon: "icon",
  };
  let wrapper: ReturnType<typeof mount<typeof WrappedFontAwesomeIcon>>;

  async function mountWrappedFontAwesomeIconComponent(options: ComponentMountingOptions<typeof WrappedFontAwesomeIcon> = {}):
  Promise<ReturnType<typeof mount<typeof WrappedFontAwesomeIcon>>> {
    return mountSuspendedComponent(WrappedFontAwesomeIcon, {
      props: defaultProps,
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountWrappedFontAwesomeIconComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Icon", () => {
    it("should pass icon when rendered.", () => {
      const icon = wrapper.findComponent<typeof WrappedFontAwesomeIcon>(".font-awesome-icon");

      expect(icon.props("icon")).toBe("icon");
    });

    it("should pass undefined size when it is not provided in props.", () => {
      const icon = wrapper.findComponent<typeof WrappedFontAwesomeIcon>(".font-awesome-icon");

      expect(icon.attributes("size")).toBeUndefined();
    });

    it("should pass provided size when it is provided in props.", async() => {
      wrapper = await mountWrappedFontAwesomeIconComponent({
        props: {
          ...defaultProps,
          size: "2x",
        },
      });
      const icon = wrapper.findComponent<typeof WrappedFontAwesomeIcon>(".font-awesome-icon");

      expect(icon.props("size")).toBe("2x");
    });

    it("should pass empty class when it is not provided in props.", () => {
      const icon = wrapper.findComponent<typeof WrappedFontAwesomeIcon>(".font-awesome-icon");

      expect(icon.attributes("class")).toBe("font-awesome-icon");
    });

    it("should pass provided class when it is provided in props.", async() => {
      wrapper = await mountWrappedFontAwesomeIconComponent({
        props: {
          ...defaultProps,
          classes: "test-class",
        },
      });
      const icon = wrapper.findComponent<typeof WrappedFontAwesomeIcon>(".font-awesome-icon");

      expect(icon.attributes("class")).toBe("font-awesome-icon test-class");
    });

    it("should set icon color to default when it is not provided in props.", () => {
      const icon = wrapper.findComponent<typeof WrappedFontAwesomeIcon>(".font-awesome-icon");

      expect(icon.attributes("style")).toBe("color: #000000;");
    });

    it("should set icon color to provided color when it is provided in props.", async() => {
      wrapper = await mountWrappedFontAwesomeIconComponent({
        props: {
          ...defaultProps,
          iconColor: "#FFFFFF",
        },
      });
      const icon = wrapper.findComponent<typeof WrappedFontAwesomeIcon>(".font-awesome-icon");

      expect(icon.attributes("style")).toBe("color: #FFFFFF;");
    });
  });
});