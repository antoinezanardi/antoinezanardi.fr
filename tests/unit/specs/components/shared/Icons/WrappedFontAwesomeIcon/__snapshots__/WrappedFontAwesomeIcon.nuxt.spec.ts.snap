// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Wrapped Font Awesome Icon Component > should match snapshot when rendered. 1`] = `"<span><font-awesome-icon-stub icon="icon" border="false" fixedwidth="false" flip="false" listitem="false" pulse="false" rotateby="false" swapopacity="false" spin="false" symbol="false" inverse="false" bounce="false" shake="false" beat="false" fade="false" beatfade="false" flash="false" spinpulse="false" spinreverse="false" widthauto="false" class="font-awesome-icon" role="button" style="color: #000000;"></font-awesome-icon-stub></span>"`;
